upp/stages/resampling.py

```diff
@@ -29,6 +30,63 @@ def safe_divide(a, b):
     return np.divide(a, b, out=np.zeros_like(a), where=b != 0)
 
 
+def subdivide_bins(bins, n=2):
+    comb_list = [np.array((bins[0],))]
+    for i in range(len(bins) - 1):
+        comb_list.append(np.linspace(bins[i], bins[i + 1], n + 1)[1:])
+    return np.concatenate(comb_list)
+
+
+def upscale_array(array, upscl, order=3, mode="nearest", nornalise=False, positive=True):
+    # upscl must be integer
+    xs = []
+    for d in array.shape:
+        n_bins = d
+        points = np.linspace(-0.5 + 1 / 2 / upscl, n_bins - 0.5 - 1 / 2 / upscl, n_bins * upscl)
+        xs.append(points)
+
+    # return the smoothed array
+    xy = np.meshgrid(*xs, indexing="ij")
+    smoothed = ndimage.map_coordinates(array, xy, order=order, mode=mode)
+    if positive:
+        smoothed[smoothed < 0] = 0
+    if nornalise:
+        smoothed = smoothed / smoothed.sum()
+    return smoothed
+
+
+def upscale_array_regionally(
+    array, upscl, regionlengthsd, order=3, mode="nearest", normalise=False
+):
+    up_array = np.empty(shape=[ds * upscl for ds in array.shape])
+    starts = [np.cumsum([0] + regionlengths)[:-1] for regionlengths in regionlengthsd]
+    starts_grid = np.meshgrid(*starts)
+    starts_grid = [starts_grid[i].flatten() for i in range(len(starts_grid))]
+    finishes = [np.cumsum(regionlengths) for regionlengths in regionlengthsd]
+    finishes_grid = np.meshgrid(*finishes)
+    finishes_grid = [finishes_grid[i].flatten() for i in range(len(finishes_grid))]
+    d = len(array.shape)
+    for i in range(len(starts_grid[0])):
+        slice_obj = []
+        slice_obj_up = []
+        for j in range(d):
+            slice_obj.append(slice(starts_grid[j][i], finishes_grid[j][i]))
+            slice_obj_up.append(slice(starts_grid[j][i] * upscl, finishes_grid[j][i] * upscl))
+        slice_obj = tuple(slice_obj)
+        slice_obj_up = tuple(slice_obj_up)
+        up_array[slice_obj_up] = upscale_array(array[slice_obj], upscl, order=order, mode=mode)
+    if normalise:
+        up_array = up_array / up_array.sum()
+    return up_array
+
+
+def regionlengthsd_from_config(config):
```

Comment:
This should be a method or property of the sampling config class

Reply:
Ok. I have changed it now